multi_camera_parallel_grabber/ex-multi_camera_parallel_grabber.cpp

```diff
@@ -0,0 +1,182 @@
+
+    //CHANGE IP addreses to those of camera's avaialible!!!!
+    const auto CAMERA0 = "192.168.0.70"; // Device in software trigger mode
+    const auto CAMERA1 = "192.168.0.71";
+    const auto CAMERA2 = "192.168.0.72";
+
+    //Add the IP of cameras to be used with hardware trigger
+    const std::vector<std::string> camera_ips = {CAMERA1,CAMERA2};
+    int num_cam_finished = camera_ips.size();
+
+    auto j_conf = json::parse(R"(
+         {
+           "ifm3d":
+           {
+             "Device":
+              {
+                "ActiveApplication": "1"
+              },
+              "Apps":
+              [
+                {
+                  "Index": "1",
+                  "TriggerMode":"2",
+                  "LogicGraph": "{\"IOMap\": {\"OUT1\": \"RFT\",\"OUT2\": \"AQUFIN\"},\"blocks\": {\"B00001\": {\"pos\": {\"x\": 200,\"y\": 200},\"properties\": {},\"type\": \"PIN_EVENT_IMAGE_ACQUISITION_FINISHED\"},\"B00002\": {\"pos\": {\"x\": 200,\"y\": 75},\"properties\": {},\"type\": \"PIN_EVENT_READY_FOR_TRIGGER\"},\"B00003\": {\"pos\": {\"x\": 600,\"y\": 75},\"properties\": {\"pulse_duration\": 0},\"type\": \"DIGITAL_OUT1\"},\"B00005\": {\"pos\": {\"x\": 600,\"y\": 200},\"properties\": {\"pulse_duration\": 0},\"type\": \"DIGITAL_OUT2\"}},\"connectors\": {\"C00000\": {\"dst\": \"B00003\",\"dstEP\": 0,\"src\": \"B00002\",\"srcEP\": 0},\"C00001\": {\"dst\": \"B00005\",\"dstEP\": 0,\"src\": \"B00001\",\"srcEP\": 0}}}",
+                  "Imager":
+                  {
+                      "ExposureTime": "1000",
+                      "Type":"under5m_moderate",
+                      "FrameRate":"20"
+                  }
+                }
+             ]
+            }
+          }
+      )");
+
+    //for configuration of required trigger on camera
+    void configuration(ifm3d::Camera::Ptr& camera, trigger_mode type)
+    {
+        try
+        {
+            auto application_id = camera->ActiveApplication();
+            j_conf["ifm3d"]["Device"]["ActiveApplication"] = std::to_string(application_id);
+            j_conf["ifm3d"]["Apps"][0]["Index"] = std::to_string(application_id);
+            j_conf["ifm3d"]["Apps"][0]["TriggerMode"] = std::to_string((int)type);
+
+            camera->FromJSON(j_conf);
+        }
+        catch (std::exception &e)
+        {
+            std::cout << e.what();
+        }
+    }
+
+    // for grabbing data when its available
+    void grabimage(ifm3d::FrameGrabber::Ptr& frame_grabber,
+            std::function<void(ifm3d::ImageBuffer::Ptr image_buffer,
+                bool in_time)> callback)
+    {
+        while(true)
+        {
+            std::this_thread::sleep_for(std::chrono::duration<int, std::milli>(1));
+            ifm3d::ImageBuffer::Ptr image_buffer = std::make_shared<ifm3d::ImageBuffer>();
+            const bool in_time = frame_grabber->WaitForFrame(image_buffer.get(), 10000);
+            callback(image_buffer,in_time);
```

Comment:
In general, it is an anti-pattern to pass a `std::shared_ptr` to a function that does not participate in the ownership of the underlying pointer. This is mostly due to performance. For each callback you are incurring a lock -> ref count increment -> unlock then at the end of the function lock -> ref count decrement -> unlock. As with my comment above, in high frequency frame grabbing loops this cost will accumulate to something non-trivial. It would be recommended to pass the naked pointer as `callback(...)` (I assume) does not participate in the ownership of the `ImageBuffer`.

Reply:
So the idea of passing the ``std::shared_ptr`` as copy was to have a clean ownership handling. If I do pass the naked pointer the ownership is not defined in a clean way. What happens if the user in the callback enqueue the ``ImageBuffer`` in an input queue of another thread for example? The callback goes out of scope the image buffer will be destroyed because the reference count is zero. We will get an dangling pointer. I agree this is not the most performant way but a trade off between performance and a ``robust`` example.

Reply:
So, I think the signature of `callback` does in fact express ownership intentions. In this case, I think you are saying that you expect `callback` to participate in the ownership of the `ImageBuffer`. As noted in my message above, I assumed it did not participate in the ownership which is why I made my comment.

I think that perhaps more realistically, the user should pass in the `ImageBuffer`. This alleviates both of my concerns: 1) too many costly heap allocations; 2) clarity on ownership. Of course, this brings up thread synchronization issues as `ImageBuffer` is not thread safe, so, perhaps also a mutex (or other lock) should be passed in as well -- better yet, an interface which wraps/subclasses (not sure if a subclass would work) the `ImageBuffer`, perhaps a `ThreadSafeImageBuffer` or something like that. This approach solves the two issues called out above *and* thread safety.

Reply:
@tpanzarella thanks for pointing out the thread safety issue of the ``ImageBuffer``